### ISSUE TITLE
Pyright flags a typing error because it is shifting the argument's type when using a Callable Protocol with keyword only args

### ISSUE THREAD

**Issue by carlos-sarmiento** [microsoft/pyright#6696]
**Describe the bug**
When using a Callable Protocol to type a keyword only function the typechecker shifts the types of the protocol's argument by one to the left. This means argument `n` uses the type of argument `n+1` instead of it's own  type. The type of the first argument is dropped and the last argument keeps its type correctly. When typechecking a function that correctly implements the protocol, pyright flags the function as not matching the definition.

**Code or Screenshots**
Fully working example in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYBQNAxFAOq4DWqGAogB4CG2FAplXoBeUWPESqpCtwDOsxizZdeCAQAoiYEuQoBKAFxUoJqABN%2BwKAH1rpbhQq31AKgA0UbiDQBGA1FkYEA8vNAAmf0CQPSgAWgA%2BKAA5MBR%2BfwA6LJoLKwB3JXRrfh4%2BfnVQvwCgkO8I6ui4xJS0zOyhKHYAN34QOBgACzYoApBmBQHe-igAI357AFdZaZntAc9vBYh%2BFBgFL2mwK0Hp2V5p%2BARBUdYiktUBay6vfyYx5VK1aeERwrRiz4CGgdABC4GYO06gOmPg6EnhkmkcgUYLAEJQKjKPk0xDIlEMxlMuRsdgcTmsrlqvkiNQ24X8wAoYG4MBiCWSqXSUCyGRyllm4J2AIe-Gs2MqNOCdPqjOZrKaHNa3Pa9EqUGAuCgqPRmK%2BPigaH4eygECQgUQ-DMngUAANZSybZgUIF%2BNwrUcoDaojaqDNBShhWUxU8Xlr-bqBPqfn60UL7kHYXQw7GUFCRVAwnCEdmpDJ5MmddCwjjtHj9EZTOZ%2BbZ7I5nO46VVUDAqfUolSAMwMpksqkAFn8q3xCpaXJ5fKsMfRga%2B1mLEqdLelks73bl-cHYGH7NHbV5HTVGpABZ2Ef4YQNRoUpvNVytck93qdLrdUA9NubPtVdXVmu1p6LS9jRvS5LWtT17RgR1UBfd0rC9IIvzpDtf2Pf8MUAw1gLNUD71tIcKGg50YFdOCIJ7KDfX9GdHjCEMQH8dCzwvaNqPjWdMyAA)

```python
from typing import Protocol


# Working Example
# ===============
class WorkingExample(Protocol):
    def __call__(*, arg1: str, arg2: str) -> None: ...

def working_example(arg1: str, arg2: str) -> None: ...

# Everything works here because both arguments are of the same type
working_example_var: WorkingExample = working_example


# Broken Example 1
# ================
class BrokenExample1(Protocol):
    def __call__(*, arg1: str, arg2: float) -> None: ...

def broken_example_1(arg1: str, arg2: float) -> None: ...

# arg1 for BrokenExample1 gets mistyped as `float` instead of `str`
broken_example_1_var: BrokenExample1 = broken_example_1


# Broken Example 2
# ================
class BrokenExample2(Protocol):
    def __call__(*, arg1: int, arg2: str, arg3: float, arg4: bool) -> None: ...

def broken_example_2(arg1: int, arg2: str, arg3: float, arg4: bool) -> None: ...

# arg1 for BrokenExample2 gets mistyped as `str` instead of `int`
# arg2 for BrokenExample2 gets mistyped as `float` instead of `str`
# arg3 for BrokenExample2 gets mistyped as `bool` instead of `float`
broken_example_2_var: BrokenExample2 = broken_example_2

```

**Full Error**
```
somePath/test.py
  somePath/test.py:31:40 - error: Expression of type "(arg1: str, arg2: float) -> None" cannot be assigned to declared type "BrokenExample1"
    Type "(arg1: str, arg2: float) -> None" cannot be assigned to type "(*, arg1: float, arg2: float) -> None"
      Keyword parameter "arg1" of type "float" cannot be assigned to type "str"
        "float" is incompatible with "str" (reportGeneralTypeIssues)
  somePath/test.py:48:40 - error: Expression of type "(arg1: int, arg2: str, arg3: float, arg4: bool) -> None" cannot be assigned to declared type "BrokenExample2"
    Type "(arg1: int, arg2: str, arg3: float, arg4: bool) -> None" cannot be assigned to type "(*, arg1: str, arg2: float, arg3: bool, arg4: bool) -> None"
      Keyword parameter "arg1" of type "str" cannot be assigned to type "int"
        "str" is incompatible with "int"
      Keyword parameter "arg2" of type "float" cannot be assigned to type "str"
        "float" is incompatible with "str" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

Command line -> v1.1.339 (also happens on latest Pylance)

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug.

Refer to [PEP 544](https://peps.python.org/pep-0544/#callback-protocols) for information about callback protocols and how they are intended to work.

The `__call__` method on a callable object is an instance method, so it must have a `self` parameter. In your code above, you've defined callback protocols with invalid `__call__` methods.

If you add the missing `self` parameter to the `__call__` methods in your protocol definitions, you'll see that the type errors go away.
